extensions/Worker.Extensions.Storage.Queues/src/QueueMessageJsonConverter.cs

```diff
@@ -0,0 +1,92 @@
+        {
+            if (reader.TokenType is not JsonTokenType.StartObject)
+            {
+                throw new JsonException("JSON payload expected to start with StartObject token.");
+            }
+
+            string messageId = String.Empty;
+            string popReceipt = String.Empty;
+            string messageText = String.Empty;
+            long dequeueCount = 1;
+            DateTime? nextVisibleOn = null;
+            DateTime? insertedOn = null;
+            DateTime? expiresOn = null;
+
+            var startDepth = reader.CurrentDepth;
+
+            while (reader.Read())
+            {
+                if (reader.TokenType is JsonTokenType.EndObject && reader.CurrentDepth == startDepth)
+                {
+                    return QueuesModelFactory.QueueMessage(
+                        messageId,
+                        popReceipt,
+                        messageText,
+                        dequeueCount,
+                        nextVisibleOn,
+                        insertedOn,
+                        expiresOn
+                    );
+                }
+
+                if (reader.TokenType is not JsonTokenType.PropertyName)
+                {
+                    continue;
+                }
+
+                var propertyName = reader.GetString();
+                reader.Read();
+
+                switch (propertyName)
+                {
+                    // We're not expecting these three values to be null. If they are, should we throw or just return an empty string?
+                    case "MessageId":
+                        messageId = reader.GetString() ?? throw new JsonException("MessageId");
+                        break;
+                    case "PopReceipt":
+                        popReceipt = reader.GetString() ?? throw new JsonException("PopReceipt");
+                        break;
+                    case "MessageText":
+                        messageText = reader.GetString() ?? throw new JsonException("MessageText");
+                        break;
+                    case "DequeueCount":
+                        dequeueCount = reader.GetInt64();
+                        break;
+                    case "NextVisibleOn":
+                        nextVisibleOn = reader.GetDateTime();
+                        break;
+                    case "InsertedOn":
+                        insertedOn = reader.GetDateTime();
+                        break;
+                    case "ExpiresOn":
+                        expiresOn = reader.GetDateTime();
+                        break;
+                    default:
+                        break;
+                }
+            }
+
+            throw new JsonException("JSON payload expected to start with EndObject token.");
+        }
+
+        public override void Write(Utf8JsonWriter writer, QueueMessage value, JsonSerializerOptions options)
+        {
+            throw new NotImplementedException();
```

Comment:
Is this the recommended exception to throw here? Or is `JsonException` more fitting?

Not saying it is, just want to understand for myself.

Reply:
I think you're right, JsonException is more fitting since it's specifically designed for handling JSON-related errors; plus it lets us catch and handle the exception specifically related to JSON processing which is better than just throwing unimplemented - thanks for catching this!